### DISCUSSION TITLE
Integrate '.setTokenAmount()' custom function for 'TokenFreezeTransaction()'

### DISCUSSION THREAD

**Discussion by Topachi** [hiero-ledger/hiero-improvement-proposals#362] [Ideas]
**hip: TBD
title: Integrate '.setTokenAmount()' custom function for 'TokenFreezeTransaction()'
author: Tomachi Anura [shibartoken@protonmail.com](mailto:shibartoken@protonmail.com)
type: Standards Track
category: Service
needs-council-approval: Yes
status: Draft
last-call-date-time:
created: 22.02.13
discussions-to:**

## Abstract

When "locking" or "freezing" liquidity related to an associated Account Id, There is no way to freeze a specific amount of tokens. The Shibar Network Team strongly believes that it would bring many advantages to have the tokenFreezeTransaction like in the sequent example:

<img width="711" alt="Screen Shot 2022-02-13 at 5 44 48 PM" src="https://user-images.githubusercontent.com/96840872/153778591-6ba9b8b7-92a7-4438-b9fb-23e77a469e4d.png">

## Motivation

When "locking" or "freezing" liquidity related to an associated Account Id, There is no way to freeze a specific amount of tokens. This could be potentially detrimental in regards of DeFi, if the same token offers many use-cases. As it stands now, Users would have to set different accounts in order to use the same Network.

## Rationale
When Alice participates to a round 1 token pre-sale, she buys 1000 tokens. According to the terms and conditions of the round 1 pre-sale, she should have those tokens frozen and unable to move them from the wallet until the term expires. However Alice would also like to buy more tokens at round 2 which are supposed to be Unfrozen tokens. Unfortunately Hedera only allows to freeze the whole token amount of the account, while it would be much more beneficial to implement a child function .setTokenAmount(Amount) to the parent TokenFreezeTransaction().

This implementation allows Alice to interact with the token if she holds more than the frozen amount. for many use cases, without limiting her whole account by keeping the whole token amount frozen. This way, she can buy another 1000 tokens on round 2, having a total balance of 2000. And she is able to use unfrozen 1000 tokens from round 2 to interact with the network

## User Stories

- As dApp, I would like to allow users to stake their tokens by freezing a partial amount of TokenX in the user wallet for a pre-determined duration

## Open Issues

No blocking issues are known at this time.
## Copyright/license

This document is licensed under the Apache License, Version 2.0 -- see [LICENSE](https://github.com/hashgraph/hedera-improvement-proposal/LICENSE) or (https://www.apache.org/licenses/LICENSE-2.0)

**Comment by AdrianKBL** [comment]
I think this could be great and would open too many different use cases as for example Non Custodial Staking, allowing the user do transaction with the amount that he hasn't staked.
